diff -crB jairhul-bdsim-c725b23739b8/include/BDSAcceleratorModel.hh jairhul-bdsim-c725b23739b8_modified/include/BDSAcceleratorModel.hh
*** jairhul-bdsim-c725b23739b8/include/BDSAcceleratorModel.hh	2024-01-29 19:21:58.000000000 +0000
--- jairhul-bdsim-c725b23739b8_modified/include/BDSAcceleratorModel.hh	2025-02-24 17:46:10.626397095 +0000
***************
*** 198,205 ****
    BDSBeamline* tunnelBeamline;            ///< Tunnel segments beam line.
    BDSBeamline* placementBeamline;         ///< Placement beam line.
    BDSBeamline* blmsBeamline;              ///< BLMs beam line.
!   
    std::vector<BDSFieldObjects*>         fields;    ///< All field objects.
    std::map<G4String, BDSRegion*>        regions;
    std::set<BDSRegion*>                  regionStorage; ///< Unique storage of regions.
    std::map<G4String, BDSApertureInfo*>  apertures; ///< All apertures.
--- 198,206 ----
    BDSBeamline* tunnelBeamline;            ///< Tunnel segments beam line.
    BDSBeamline* placementBeamline;         ///< Placement beam line.
    BDSBeamline* blmsBeamline;              ///< BLMs beam line.
!   public:
    std::vector<BDSFieldObjects*>         fields;    ///< All field objects.
+   private:
    std::map<G4String, BDSRegion*>        regions;
    std::set<BDSRegion*>                  regionStorage; ///< Unique storage of regions.
    std::map<G4String, BDSApertureInfo*>  apertures; ///< All apertures.
Only in jairhul-bdsim-c725b23739b8_modified/include: BDSAcceleratorModel.hh.rej
diff -crB jairhul-bdsim-c725b23739b8/include/BDSFieldBuilder.hh jairhul-bdsim-c725b23739b8_modified/include/BDSFieldBuilder.hh
*** jairhul-bdsim-c725b23739b8/include/BDSFieldBuilder.hh	2024-01-29 19:21:58.000000000 +0000
--- jairhul-bdsim-c725b23739b8_modified/include/BDSFieldBuilder.hh	2025-02-24 17:46:10.688397181 +0000
***************
*** 84,93 ****
  
    /// Singleton instance.
    static BDSFieldBuilder* instance;
!   
    /// @{ Register of components to build.
    std::vector<const BDSFieldInfo*> infos;
    std::vector<std::vector<G4LogicalVolume*> > lvs;
    std::vector<G4bool>              propagators;
    /// @}
  
--- 84,94 ----
  
    /// Singleton instance.
    static BDSFieldBuilder* instance;
!   public:
    /// @{ Register of components to build.
    std::vector<const BDSFieldInfo*> infos;
    std::vector<std::vector<G4LogicalVolume*> > lvs;
+   private:
    std::vector<G4bool>              propagators;
    /// @}
  
Only in jairhul-bdsim-c725b23739b8_modified/include: BDSFieldBuilder.hh.rej
diff -crB jairhul-bdsim-c725b23739b8/include/BDSOutput.hh jairhul-bdsim-c725b23739b8_modified/include/BDSOutput.hh
*** jairhul-bdsim-c725b23739b8/include/BDSOutput.hh	2024-01-29 19:21:58.000000000 +0000
--- jairhul-bdsim-c725b23739b8_modified/include/BDSOutput.hh	2025-04-10 10:41:44.962956530 +0100
***************
*** 157,162 ****
--- 157,164 ----
    /// Feedback for protected names.
    static void PrintProtectedNames(std::ostream& out);
  
+   using BDSOutputStructures::samplerTrees;
+   bool storeOutput = true;
  protected:
    /// Get the next file name based on the base file name and the accrued number of files.
    G4String GetNextFileName();
Only in jairhul-bdsim-c725b23739b8_modified/include: BDSOutput.hh.rej
diff -crB jairhul-bdsim-c725b23739b8/include/BDSOutputROOTEventSampler.hh jairhul-bdsim-c725b23739b8_modified/include/BDSOutputROOTEventSampler.hh
*** jairhul-bdsim-c725b23739b8/include/BDSOutputROOTEventSampler.hh	2024-01-29 19:21:58.000000000 +0000
--- jairhul-bdsim-c725b23739b8_modified/include/BDSOutputROOTEventSampler.hh	2025-05-29 11:29:28.682990812 +0100
***************
*** 50,55 ****
--- 50,57 ----
    std::vector<U>     energy;
    std::vector<U>     x;
    std::vector<U>     y;
+   std::vector<U>     xcache;
+   std::vector<U>     ycache;
    U                  z;
    std::vector<U>     xp;
    std::vector<U>     yp;
***************
*** 96,101 ****
--- 98,105 ----
    explicit BDSOutputROOTEventSampler(std::string samplerNameIn);
    virtual ~BDSOutputROOTEventSampler();
  #ifndef __ROOTBUILD__
+ 
+   void FillSimple(const BDSHitSampler* hit);
    void Fill(const BDSHitSampler* hit,
  	    G4bool storeMass          = false,
  	    G4bool storeCharge        = false,
***************
*** 132,137 ****
--- 136,142 ----
    void SetBranchAddress(TTree *);
    virtual void Flush();  ///< Clean Sampler
  
+   void FlushCache();
    void FlushLocal(); ///< Actual flush that is non virtual function to use in constructor.
  
    static BDSOutputROOTParticleData* particleTable;
Only in jairhul-bdsim-c725b23739b8_modified/include: BDSOutputROOTEventSampler.hh.rej
diff -crB jairhul-bdsim-c725b23739b8/include/BDSOutputStructures.hh jairhul-bdsim-c725b23739b8_modified/include/BDSOutputStructures.hh
*** jairhul-bdsim-c725b23739b8/include/BDSOutputStructures.hh	2024-01-29 19:21:58.000000000 +0000
--- jairhul-bdsim-c725b23739b8_modified/include/BDSOutputStructures.hh	2025-02-24 17:46:10.692397187 +0000
***************
*** 144,154 ****
--- 144,156 ----
  #endif
    BDSOutputROOTEventCoords* primaryGlobal; /// Global coordinates of primary particle.
    /// Sampler structures.
+ public:
  #ifdef __ROOTDOUBLE__
    std::vector<BDSOutputROOTEventSampler<double>*> samplerTrees;
  #else
    std::vector<BDSOutputROOTEventSampler<float>*> samplerTrees;
  #endif
+ protected:
    std::vector<std::string> samplerNames; ///< Sampler names to use.
    std::vector<BDSOutputROOTEventSamplerC*> samplerCTrees;
    std::vector<BDSOutputROOTEventSamplerS*> samplerSTrees;
Only in jairhul-bdsim-c725b23739b8_modified/include: BDSOutputStructures.hh.rej
diff -crB jairhul-bdsim-c725b23739b8/include/BDSRandom.hh jairhul-bdsim-c725b23739b8_modified/include/BDSRandom.hh
*** jairhul-bdsim-c725b23739b8/include/BDSRandom.hh	2024-01-29 19:21:58.000000000 +0000
--- jairhul-bdsim-c725b23739b8_modified/include/BDSRandom.hh	2025-02-24 17:46:10.692397187 +0000
***************
*** 46,51 ****
--- 46,53 ----
  
    /// Set the initial seed. Takes the seed number from
    /// BDSGlobalConstants - if negative uses the time.
+   void SetSeed(long seed);
+ 
    void SetSeed();
  
    /// Print out seed state to G4cout.
Only in jairhul-bdsim-c725b23739b8_modified/include: BDSRandom.hh.rej
diff -crB jairhul-bdsim-c725b23739b8/src/BDSEventAction.cc jairhul-bdsim-c725b23739b8_modified/src/BDSEventAction.cc
*** jairhul-bdsim-c725b23739b8/src/BDSEventAction.cc	2024-01-29 19:21:58.000000000 +0000
--- jairhul-bdsim-c725b23739b8_modified/src/BDSEventAction.cc	2025-02-24 17:46:10.694397190 +0000
***************
*** 184,191 ****
    G4int currentEventID = evt->GetEventID();
    BDSSDTerminator::eventNumber = currentEventID; // update static member of terminator
    eventInfo->SetIndex(currentEventID);
!   if (currentEventID % printModulo == 0)
!     {G4cout << "---> Begin of event: " << currentEventID << G4endl;}
    if (verboseEventBDSIM) // always print this out
      {G4cout << __METHOD_NAME__ << "event #" << currentEventID << G4endl;}
  
--- 184,192 ----
    G4int currentEventID = evt->GetEventID();
    BDSSDTerminator::eventNumber = currentEventID; // update static member of terminator
    eventInfo->SetIndex(currentEventID);
!   //CERN unnecessary output
! //  if (currentEventID % printModulo == 0)
! //    {G4cout << "---> Begin of event: " << currentEventID << G4endl;}
    if (verboseEventBDSIM) // always print this out
      {G4cout << __METHOD_NAME__ << "event #" << currentEventID << G4endl;}
  
Only in jairhul-bdsim-c725b23739b8_modified/src: BDSEventAction.cc.rej
diff -crB jairhul-bdsim-c725b23739b8/src/BDSOutput.cc jairhul-bdsim-c725b23739b8_modified/src/BDSOutput.cc
*** jairhul-bdsim-c725b23739b8/src/BDSOutput.cc	2024-01-29 19:21:58.000000000 +0000
--- jairhul-bdsim-c725b23739b8_modified/src/BDSOutput.cc	2025-05-29 12:02:49.001654564 +0100
***************
*** 329,374 ****
    energyWorldExit              = 0;
    energyWorldExitKinetic       = 0;
    nCollimatorsInteracted       = 0;
!   
!   if (vertex && storePrimaries)
!     {FillPrimary(vertex, turnsTaken);}
    FillSamplerHitsVector(samplerHitsPlane);
!   FillSamplerCylinderHitsVector(samplerHitsCylinder);
!   FillSamplerSphereHitsVector(samplerHitsSphere);
!   if (samplerHitsLink)
!     {FillSamplerHitsLink(samplerHitsLink);}
!   if (energyLoss)
!     {FillEnergyLoss(energyLoss,        BDSOutput::LossType::energy);}
!   if (energyLossFull)
!     {FillEnergyLoss(energyLossFull,    BDSOutput::LossType::energy);}
!   if (energyLossVacuum)
!     {FillEnergyLoss(energyLossVacuum,  BDSOutput::LossType::vacuum);}
!   if (energyLossTunnel)
!     {FillEnergyLoss(energyLossTunnel,  BDSOutput::LossType::tunnel);}
!   if (energyLossWorld)
!     {FillEnergyLoss(energyLossWorld,   BDSOutput::LossType::world);}
!   if (worldExitHits)
!     {FillEnergyLoss(worldExitHits,     BDSOutput::LossType::worldexit);}
!   if (energyLossWorldContents)
!     {FillEnergyLoss(energyLossWorldContents, BDSOutput::LossType::worldcontents);}
!   FillPrimaryHit(primaryHits);
!   FillPrimaryLoss(primaryLosses);
!   if (trajectories)
!     {FillTrajectories(trajectories);}
!   if (collimatorHits)
!     {FillCollimatorHits(collimatorHits, primaryLosses);}
!   if (apertureImpacts)
!     {FillApertureImpacts(apertureImpactHits);}
!   FillScorerHits(scorerHits); // map always exists
! 
!   // we do this after energy loss and collimator hits as the energy loss
!   // is integrated for putting in event info and the number of collimators
!   // interacted with counted
!   if (info)
!     {FillEventInfo(info);}
    
!   WriteFileEventLevel();
!   ClearStructuresEventLevel();
  }
  
  void BDSOutput::CloseAndOpenNewFile()
--- 329,378 ----
    energyWorldExit              = 0;
    energyWorldExitKinetic       = 0;
    nCollimatorsInteracted       = 0;
! 
    FillSamplerHitsVector(samplerHitsPlane);
!  //CERN writing to file is such a drag when we only need the primary positions on the samplers
!   if(storeOutput){
!       
!     if (vertex && storePrimaries)
!       {FillPrimary(vertex, turnsTaken);}
!     FillSamplerCylinderHitsVector(samplerHitsCylinder);
!     FillSamplerSphereHitsVector(samplerHitsSphere);
!     if (samplerHitsLink)
!       {FillSamplerHitsLink(samplerHitsLink);}
!     if (energyLoss)
!       {FillEnergyLoss(energyLoss,        BDSOutput::LossType::energy);}
!     if (energyLossFull)
!       {FillEnergyLoss(energyLossFull,    BDSOutput::LossType::energy);}
!     if (energyLossVacuum)
!       {FillEnergyLoss(energyLossVacuum,  BDSOutput::LossType::vacuum);}
!     if (energyLossTunnel)
!       {FillEnergyLoss(energyLossTunnel,  BDSOutput::LossType::tunnel);}
!     if (energyLossWorld)
!       {FillEnergyLoss(energyLossWorld,   BDSOutput::LossType::world);}
!     if (worldExitHits)
!       {FillEnergyLoss(worldExitHits,     BDSOutput::LossType::worldexit);}
!     if (energyLossWorldContents)
!       {FillEnergyLoss(energyLossWorldContents, BDSOutput::LossType::worldcontents);}
!     FillPrimaryHit(primaryHits);
!     FillPrimaryLoss(primaryLosses);
!     if (trajectories)
!       {FillTrajectories(trajectories);}
!     if (collimatorHits)
!       {FillCollimatorHits(collimatorHits, primaryLosses);}
!     if (apertureImpacts)
!       {FillApertureImpacts(apertureImpactHits);}
!     FillScorerHits(scorerHits); // map always exists
    
!     // we do this after energy loss and collimator hits as the energy loss
!     // is integrated for putting in event info and the number of collimators
!     // interacted with counted
!     if (info)
!       {FillEventInfo(info);}
!     
!     WriteFileEventLevel();
!     ClearStructuresEventLevel();
!   }
  }
  
  void BDSOutput::CloseAndOpenNewFile()
***************
*** 386,395 ****
                          unsigned long long int nEventsDistrFileSkippedIn,
                          unsigned int distrFileLoopNTimesIn)
  {
!   FillRunInfoAndUpdateHeader(info, nOriginalEventsIn, nEventsRequestedIn, nEventsInOriginalDistrFileIn, nEventsDistrFileSkippedIn, distrFileLoopNTimesIn);
!   WriteFileRunLevel();
!   WriteHeaderEndOfFile();
!   ClearStructuresRunLevel();
  }
  
  G4bool BDSOutput::InvalidSamplerName(const G4String& samplerName)
--- 390,401 ----
                          unsigned long long int nEventsDistrFileSkippedIn,
                          unsigned int distrFileLoopNTimesIn)
  {
!   if(storeOutput){
!     FillRunInfoAndUpdateHeader(info, nOriginalEventsIn, nEventsRequestedIn, nEventsInOriginalDistrFileIn, nEventsDistrFileSkippedIn, distrFileLoopNTimesIn);
!     WriteFileRunLevel();
!     WriteHeaderEndOfFile();
!     ClearStructuresRunLevel();
!   }
  }
  
  G4bool BDSOutput::InvalidSamplerName(const G4String& samplerName)
***************
*** 719,727 ****
            const BDSHitSampler* hit = (*hc)[i];
            G4int samplerID = hit->samplerID;
            G4int samplerVectorIndex = samplerIDToIndexPlane[samplerID];
!           samplerTrees[samplerVectorIndex]->Fill(hit, storeSamplerMass, storeSamplerCharge,
!                                                  storeSamplerPolarCoords, storeSamplerIon,
!                                                  storeSamplerRigidity, storeSamplerKineticEnergy);
          }
      }
    // extra information - do only once at the end
--- 725,741 ----
            const BDSHitSampler* hit = (*hc)[i];
            G4int samplerID = hit->samplerID;
            G4int samplerVectorIndex = samplerIDToIndexPlane[samplerID];
! 
!           if(storeOutput)
!           {
!               samplerTrees[samplerVectorIndex]->Fill(hit, storeSamplerMass, storeSamplerCharge,
!                                                      storeSamplerPolarCoords, storeSamplerIon,
!                                                      storeSamplerRigidity, storeSamplerKineticEnergy);
!           }
!           else
!           {
!             samplerTrees[samplerVectorIndex]->FillSimple(hit);
!           }
          }
      }
    // extra information - do only once at the end
Only in jairhul-bdsim-c725b23739b8_modified/src: BDSOutput.cc.rej
diff -crB jairhul-bdsim-c725b23739b8/src/BDSOutputROOTEventSampler.cc jairhul-bdsim-c725b23739b8_modified/src/BDSOutputROOTEventSampler.cc
*** jairhul-bdsim-c725b23739b8/src/BDSOutputROOTEventSampler.cc	2024-01-29 19:21:58.000000000 +0000
--- jairhul-bdsim-c725b23739b8_modified/src/BDSOutputROOTEventSampler.cc	2025-05-29 11:28:43.213643986 +0100
***************
*** 55,60 ****
--- 55,69 ----
  {;}
  
  #ifndef __ROOTBUILD__
+ 
+ 
+ template <class U>
+ void BDSOutputROOTEventSampler<U>::FillSimple(const BDSHitSampler* hit){
+ 
+   xcache.push_back((U) (hit->coords.x / CLHEP::m));
+   ycache.push_back((U) (hit->coords.y / CLHEP::m));
+ }
+ 
  template <class U>
  void BDSOutputROOTEventSampler<U>::Fill(const BDSHitSampler* hit,
  					G4bool storeMass,
***************
*** 72,77 ****
--- 81,89 ----
    energy.push_back((U) (hit->coords.totalEnergy / CLHEP::GeV));
    x.push_back((U) (hit->coords.x / CLHEP::m));
    y.push_back((U) (hit->coords.y / CLHEP::m));
+   
+   xcache.push_back((U) (hit->coords.x / CLHEP::m));
+   ycache.push_back((U) (hit->coords.y / CLHEP::m));
  
    xp.push_back((U) (hit->coords.xp));
    yp.push_back((U) (hit->coords.yp));
***************
*** 300,305 ****
--- 312,324 ----
  }
  
  template <class U>
+ void BDSOutputROOTEventSampler<U>::FlushCache()
+ {
+   xcache.clear();
+   ycache.clear();
+ }
+ 
+ template <class U>
  std::vector<U> BDSOutputROOTEventSampler<U>::getKineticEnergy()
  {
    std::vector<U> result((unsigned long)n);
Only in jairhul-bdsim-c725b23739b8_modified/src: BDSOutputROOTEventSampler.cc.rej
diff -crB jairhul-bdsim-c725b23739b8/src/BDSRandom.cc jairhul-bdsim-c725b23739b8_modified/src/BDSRandom.cc
*** jairhul-bdsim-c725b23739b8/src/BDSRandom.cc	2024-01-29 19:21:58.000000000 +0000
--- jairhul-bdsim-c725b23739b8_modified/src/BDSRandom.cc	2025-02-24 17:46:10.699397197 +0000
***************
*** 89,94 ****
--- 89,103 ----
      }
  }
  
+ void BDSRandom::SetSeed(long seed)
+ {
+   CLHEP::HepRandom::setTheSeed(seed);
+ 
+   // feedback - get the seed from the generator itself (ensures set correctly)
+   G4cout << __METHOD_NAME__ << "Random number generator's seed = "
+ 	 << CLHEP::HepRandom::getTheSeed() << G4endl;
+ }
+ 
  void BDSRandom::SetSeed()
  {
  #ifdef BDSDEBUG
Only in jairhul-bdsim-c725b23739b8_modified/src: BDSRandom.cc.rej
